In [19]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px

# Лабораторная работа №1
## Задание 2
### {Фамилия Имя}, {Номер группы}, Вариант {Номер варианта}, ({Дата})

### Задание

По данным выборки ($n=150$) требуется:
1. Составить статистическое распределение выборки, предварительно записав дискретный вариационный ряд.
1. Составить ряд распределения относительных частот.
2. Построить полигон частот.
3. Составить эмпирическую функцию распределения.
4. Построить график эмпирической функции распределения.
5. Найти основные числовые характеристики вариационного ряда:
	- Выборочное среднее - $\overline x_в$
	- Выборочную дисперсию - $S^2$
	- Стандартное отклонение (Среднеквадратическое отклонение) - $S$
	- Коэффициент вариации - $CV$
7. Пояснить смысл полученных результатов.

### Данные

In [20]:
lb_author = 'Кондрашин Тимофей'
lb_num = 1
lb_excercise_num = 2
lb_variant = 1
print(f'Фамилия Имя: {lb_author}\nНомер лабораторной работы: {lb_num}\nНомер задания лабораторной работы: {lb_excercise_num}\nВариант лабораторной работы: {lb_variant}')

Фамилия Имя: Кондрашин Тимофей
Номер лабораторной работы: 1
Номер задания лабораторной работы: 2
Вариант лабораторной работы: 1


In [21]:
data = pd.DataFrame(
    [2, 3, 4, 0, 2, 5, 3, 3, 3, 3, 2, 0, 6, 2, 3, 1, 5, 2, 4, 2, 4, 3, 1, 2, 3, 2, 2, 2, 3, 4, 1, 6, 2, 3, 3, 2, 0, 6,
     2, 5, 0, 2, 4, 3, 2, 6, 0, 2, 5, 3, 3, 3, 5, 4, 3, 1, 4, 5, 4, 3, 2, 1, 2, 3, 2, 2, 0, 2, 5, 3, 3, 1, 6, 2, 4, 6,
     0, 4, 3, 1, 4, 5, 2, 4, 2, 4, 3, 2, 3, 4, 5, 2, 0, 2, 4, 3, 2, 1, 6, 2, 3, 3, 2, 0, 6, 0, 2, 5, 3, 3, 3, 1, 2, 3,
     2, 2, 5, 2, 3, 4, 5, 4, 2, 2, 4, 2, 4, 3, 4, 3, 1, 4, 5, 2, 5, 3, 3, 3, 2, 0, 6, 1, 6, 2, 0, 2, 4, 3, 2, 0])
data.T

0    1    2    3    4    5    6    7    8    9    ...  140  141  142  143  \
0    2    3    4    0    2    5    3    3    3    3  ...    6    1    6    2   

   144  145  146  147  148  149  
0    0    2    4    3    2    0  

[1 rows x 150 columns]

### Всего элементов ряда

In [22]:
data_len = len(data)
data_len

150

## Пункт 1
**Составить статистическое распределение выборки, предварительно записав дискретный вариационный ряд.**

### Вариационный ряд

In [23]:
data_sort = data.sort_values(by=[0])
data_sort = data_sort[0].tolist()
data_sort

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 4,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 5,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6,
 6]

## Пункт 2
**Составить ряд распределения относительных частот.**

### Варианты, $x_i$

In [24]:
unique = data[0].unique().tolist()
unique.sort()
unique

[0, 1, 2, 3, 4, 5, 6]

### Частота, $n_i$

In [25]:
freq = []
for x in unique:
    freq.append(data_sort.count(x))
freq = pd.DataFrame(freq).T
freq = freq.rename(columns={0: '0', 1: '1', 2: '2', 3: '3', 4: '4', 5: '5', 6: '6'})
freq

0   1   2   3   4   5   6
0  13  11  43  37  22  14  10

### Относительная частота, $w_i$

In [26]:
rel_freq = []
for x in unique:
    rel_freq.append(data_sort.count(x) / data_len)
rel_freq = pd.DataFrame(rel_freq).T
rel_freq = rel_freq.rename(columns={0: '0', 1: '1', 2: '2', 3: '3', 4: '4', 5: '5', 6: '6'})
rel_freq

0         1         2         3         4         5         6
0  0.086667  0.073333  0.286667  0.246667  0.146667  0.093333  0.066667

## Пункт 3
**Построить полигон частот.**

### Полигон частот вариационного ряда

In [27]:
px.line(x=unique, y=freq.iloc[0], markers=True, title='£')

### Полигон относительных частот вариационного ряда

In [28]:
px.line(x=unique, y=rel_freq.iloc[0], markers=True, title='Полигон относительных частот вариационного ряда')

## Пункт 4
**Составить эмпирическую функцию распределения.**

### Эмпирическая функция распределения $F^*$

In [29]:
emp_func = []
for i, x in enumerate(unique):
    emp_func.append(sum(rel_freq.values.tolist()[0][:i]))
emp_func = pd.DataFrame(emp_func).T
emp_func = emp_func.rename(columns={0: '0', 1: '1', 2: '2', 3: '3', 4: '4', 5: '5', 6: '6'})
emp_func

0         1     2         3         4     5         6
0  0.0  0.086667  0.16  0.446667  0.693333  0.84  0.933333

$$
F^* (x) =
\begin{cases}
0, && x \le 0 \\
0.086667, && 0 < x \le 1 \\
0.16, && 1 < x \le 2 \\
0.446667, && 2 < x \le 3 \\
0.693333, && 3 < x \le 4 \\
0.84, && 4 < x \le 5 \\
0.933333, && 5 < x \le 6 \\
1, && x > 6
\end{cases}
$$

## Пункт 5
**Построить график эмпирической функции распределения.**

### График эмпирической функции распределения

In [30]:
px.ecdf(x=emp_func.columns.values.tolist(), y=emp_func.values.tolist()[0], markers=True, ecdfmode="standard",
        title='График эмпирической функции распределения')

## Пункт 6
**Найти основные числовые характеристики вариационного ряда**

### Выборочное среднее

$$\overline x = \frac1n \sum^n_{i=1} x_i$$

In [31]:
mean = np.mean(data)
mean[0]

/Users/cosmdandy/anaconda3/envs/Experiments/lib/python3.10/site-packages/numpy/core/fromnumeric.py:3462: FutureWarning:

In a future version, DataFrame.mean(axis=None) will return a scalar mean over the entire DataFrame. To retain the old behavior, use 'frame.mean(axis=0)' or just 'frame.mean()'



2.84

### Выборочная дисперсия

$$s^2_x = \frac{1}{n-1}\sum^n_{i=1} (x_i - \overline x)^2$$

In [32]:
var = np.var(data)
var[0]

2.4544

### Среднеквадратическое отклонение

$$s = \sqrt{s^2}$$

In [33]:
std = np.std(data)
std[0]

1.5666524822052912

### Коэффицент вариации

$$CV = \frac{s}{\overline x} \cdot 100$$

In [34]:
cv = (std / mean) * 100
cv[0]

55.163819795960954

## Пункт 7
**Пояснить смысл полученных результатов.**

$$\overline x = 2.84$$
$$s^2_x = 2.4544$$
$$s = 1.5666524822052912$$
$$CV = 55.163819795960954$$

In [35]:
answer = pd.DataFrame()
answer['Фамилия Имя'] = [lb_author]
answer['Номер лабораторной работы'] = [lb_num]
answer['Номер задания лабораторной работы'] = [lb_excercise_num]
answer['Вариант лабораторной работы'] = [lb_variant]
answer['Вариационный ряд'] = [str(data_sort)]
answer['Варианты'] = [str(unique)]
answer['Частота'] = [str(freq.iloc[0].tolist())]
answer['Относительная частота'] = [str(rel_freq.iloc[0].tolist())]
answer['Эмпирическая функция распределения'] = [str(emp_func.iloc[0].tolist())]
answer['Выборочное среднее'] = [mean[0]]
answer['Выборочная дисперсия'] = [var[0]]
answer['Среднеквадратическое отклонение'] = [std[0]]
answer['Коэффицент вариации'] = [cv[0]]
answer.to_csv('Кондрашин_ЛБ1_Задание_2.csv', index=False)
answer

Фамилия Имя  Номер лабораторной работы  \
0  Кондрашин Тимофей                          1   

   Номер задания лабораторной работы  Вариант лабораторной работы  \
0                                  2                            1   

                                    Вариационный ряд               Варианты  \
0  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, ...  [0, 1, 2, 3, 4, 5, 6]   

                        Частота  \
0  [13, 11, 43, 37, 22, 14, 10]   

                               Относительная частота  \
0  [0.08666666666666667, 0.07333333333333333, 0.2...   

                  Эмпирическая функция распределения  Выборочное среднее  \
0  [0.0, 0.08666666666666667, 0.16, 0.44666666666...                2.84   

   Выборочная дисперсия  Среднеквадратическое отклонение  Коэффицент вариации  
0                2.4544                         1.566652             55.16382